Fix package deps

rprije opened the pull request:
Fixes obsidiansystems/obelisk#795

The problem in the above issue is that the `ghci` launched by the `ob` command suite by default exposes all packages in the project's GHC package database. The full list of packages can be seen by using `ghc-pkg list` in a nix shell within an initialised Obelisk project. This list contains both `jsaddle-dom` and `ghcjs-dom` giving rise to the error seen in obsidiansystems/obelisk#795 as well as other problems such as concealing a failure to include needed modules in a project's `.cabal` file.

`cabal new-repl` deals with this problem by passing `-hide-all-packages` to `ghci`. However, this means that all package dependencies must be explicitly exposed with `-package-id`  arguments which `cabal` takes care of.

This PR imitates `cabal`'s behaviour by adding `-hide-all-packages`, extracting build dependencies from the `.cabal` files, and adding them as `-package-id` arguments.

`ob` seems to overlap with `cabal` in a lot of its scope. I wonder if a better approach than further work akin to what I'm doing here might be to backport to `cabal` the things the `ob` command suite does better than `cabal`, such as the ability to load multiple modules, and move to using that instead?


I have:

  - [X] Based work on latest `develop` branch
  - [X] Looked for lint in my changes with `hlint .` (lint found code you did not write can be left alone)
  - [X] Run the test suite: `$(nix-build -A selftest --no-out-link)`
  - [ ] (Optional) Run CI tests locally: `nix-build release.nix -A build.x86_64-linux --no-out-link` (or `x86_64-darwin` on macOS)

3noch commented:
Awesome!

rprije commented:
I tried this out against my personal project. It fixed the problem I was having with my `GHCJS.DOM.Document` module import and additionally notified me of some packages I'd forgotten to add to my `.cabal` file that were previously being concealed by this problem. I haven't noticed any new problems introduced. This PR is now good to review.

Ericson2314 commented:
> `ob` seems to overlap with `cabal` in a lot of its scope. I wonder if a better approach than further work akin to what I'm doing here might be to backport to `cabal` the things the `ob` command suite does better than `cabal`, such as the ability to load multiple modules, and move to using that instead?

There was a GSOC this summer (which I helped mentor) for getting support for multiple components in GHC itself: https://summerofcode.withgoogle.com/projects/#5269390116782080. Folow-up work for Cabal and HIE has been started, after which I suppose we can drop what we have in `ob`.

rprije commented:
Updated this PR with a fix and rebased to the latest `develop` branch.

Details: After doing some updates of NixOS and Obelisk I discovered there were some problems with this PR. It was failing to load the GHC package DB when the system GHC was for a different version of GHC to the project package DB. This has now been fixed by cleaning up the code that handles loading the package db to be more consistent in getting the correct `ghc` and `ghc-pkg` programs and package db for the project from Nix.

rprije commented:
Is there anything I can do to help get this merged? I'd like to avoid resolving merge conflicts a 4th time.

ali-abrar commented:
Thanks! Apologies for the delay.
